herd/valconstraint.ml

```diff
@@ -219,18 +240,81 @@ and type state = A.state =
         let v = subst_atom m v
         and w = subst_atom m w in
         if V.compare v w = 0 then k else Assign (v,Atom w)::k
+    | Assign (v,Deref w) ->
+        let v = subst_atom m v
+        and w = subst_atom m w in
+        if V.compare v w = 0 then k else Assign (v,Deref w)::k
+    | Assign (v,Address w) ->
+        let v = subst_atom m v
+        and w = subst_atom m w in
+        if V.compare v w = 0 then k else Assign (v,Address w)::k
     | Assign (v,e) ->
         let v = subst_atom m v
         and e = subst_expr m e in
         Assign (v,e)::k
-    | Unroll _ -> cn::k
+    | AssignAddr (v,e) ->
+        let v = subst_atom m v
+        and e = subst_expr m e in
+        AssignAddr (v,e)::k
+    | Unroll _ | Inv _ -> cn::k
 
     let subst_cns soln cns = List.fold_right (subst_cn soln) cns []
 
+    let match_addr v1 = function
+      | Inv (AssignAddr (v2, Atom v)) when v1 = v2-> Some v
+      | _ -> None
+
+    let match_loc v1 = function
+      | Inv (AssignAddr(v, Atom v2)) when v1 = v2-> Some v
+      | _ -> None
+
+    let simplify_addr_in_cnstrnt cs c =
+
+      let find_addr a =
+        (* Subst a symbol for a numeric address *)
+        let vs = List.filter_map (match_addr a) cs in
+        if List.length vs == 0 then begin
+          (* If we can't find the address for a, and it is concrete,
+           * or symbolic (but not known) *)
+          if V.is_concrete a || V.is_symbol a then
+            (false, Atom a)
+          else
+            (false, Address a)
+        end else
+          (true, Atom (List.hd vs)) in
```

Comment:
when we collect addresses in the initial state we use the same `Set` library to collect unique assignments to addresses so we should never have more than 1 numeric address for a location - the `_::_::_` should never occur.

More generally anything other than `[]` or `[v]` should fail and so I can add a failure case for just `_ -> Warn.fatal "Multiple ... "`

Perhaps a better data structure would work here, using `List.find_map` to return the first element that succeeds, or returning `None`.

What do you think?

Reply:
more specifically something that returns either nothing or one_element is the safest like you say, hmm. I'll do what your suggestion says

Reply:
Will see that once I can look at the new code.

Reply:
Well I am not that sure since your solution, although a bit less efficient performs a sanity check.

Anyhow, there is such a function: [`List.find_opt`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/List.html#VALfind_opt) :
```
let rec find_opt p = function
| [] -> None
| x::xs ->
    if p x then Some x else find_opt p xs
```
There is also [`List.find`](https://caml.inria.fr/pub/docs/manual-ocaml/libref/List.html#VALfind) that raises exception `Not_found` when search fail.